net/queue/ff.go

```diff
@@ -0,0 +1,256 @@
+	return len(q.index)
+}
+
+/* Call queue implementation */
+
+var _ CallQueue = (*ffQueue)(nil)
+
+type ffQueue struct {
+	peers    map[peer.ID]*peerQueue
+	inflight map[uint64]struct{}
+	poll     time.Duration
+	timeout  time.Duration
+	ctx      context.Context
+	mx       sync.Mutex
+}
+
+// Fair FIFO-queue with isolated per-peer processing and adaptive invocation rate.
+// Queue is polled with specified frequency and all scheduled calls expected to be
+// spawned within given timeout. At every moment only one call for the peer/thread
+// pair exists in the queue. Scheduled operations could be replaced with a new ones
+// based on the priority value (new higher-priority call replaces waiting one).
+func NewFFQueue(
+	ctx context.Context,
+	pollInterval time.Duration,
+	spawnTimeout time.Duration,
+) *ffQueue {
+	return &ffQueue{
+		ctx:      ctx,
+		poll:     pollInterval,
+		timeout:  spawnTimeout,
+		inflight: make(map[uint64]struct{}),
+		peers:    make(map[peer.ID]*peerQueue),
+	}
+}
+
+func (q *ffQueue) Schedule(
+	pid peer.ID,
+	tid thread.ID,
+	priority int,
+	call PeerCall,
+) bool {
+	h := hash(pid, tid)
+	q.mx.Lock()
+	if _, inflight := q.inflight[h]; inflight {
+		q.mx.Unlock()
+		log.Debugf("skip call to [%s/%s]: in-flight", pid, tid)
+		return false
+	}
+	pq, exist := q.peers[pid]
+	if !exist {
+		pq = newPeerQueue()
+		q.peers[pid] = pq
+		go q.pollQueue(pid, pq)
+	}
+	q.mx.Unlock()
+
+	pq.Lock()
+	defer pq.Unlock()
+	return pq.Add(tid, call, priority)
+}
+
+func (q *ffQueue) Call(
+	pid peer.ID,
+	tid thread.ID,
+	call PeerCall,
+) error {
+	h := hash(pid, tid)
+	q.mx.Lock()
+	pq, exist := q.peers[pid]
+	q.inflight[h] = struct{}{}
+	q.mx.Unlock()
+
+	if exist {
+		pq.Lock()
+		removed := pq.Remove(tid)
+		pq.Unlock()
+		if removed {
+			log.Debugf("deschedule call to [%s/%s]: directly invoked", pid, tid)
+		}
+	}
+
+	err := call(q.ctx, pid, tid)
+	q.mx.Lock()
+	delete(q.inflight, h)
+	q.mx.Unlock()
+	return err
+}
+
+func (q *ffQueue) pollQueue(pid peer.ID, pq *peerQueue) {
+	var tick = time.NewTicker(q.poll)
+
+	for {
+		select {
+		case <-q.ctx.Done():
+			tick.Stop()
+			return
+
+		case <-tick.C:
+			pq.Lock()
+			var deadline = time.Now().Add(-q.timeout).Unix()
+			for waiting := pq.Size(); waiting > 0; waiting-- {
+				call, tid, created, ok := pq.Pop()
+				if !ok {
+					break
+				}
+
+				go func() {
+					var h = hash(pid, tid)
+
+					// set in-flight status
+					q.mx.Lock()
+					q.inflight[h] = struct{}{}
+					q.mx.Unlock()
+
+					// make a call
+					if err := call(q.ctx, pid, tid); err != nil {
+						log.Errorf("call to [%s/%s] failed: %v", pid, tid, err)
+					}
+
+					// clear in-flight status
+					q.mx.Lock()
+					delete(q.inflight, h)
+					q.mx.Unlock()
+				}()
+
+				// spawn all overdue calls and a few ones with coming deadline
+				if remainIters := int(float64(created-deadline) / q.poll.Seconds()); remainIters > 0 &&
+					rand.Float64() > math.Sqrt(3*float64(waiting))/float64(remainIters) {
```

Comment:
As the reason of the issue is contention on the thread's head-keys, I believe that optimal solution depends on the intensity of writes to the logs. While contention is relatively low, I'd rather use optimistic approach with retried transactions. In case of unusually intensive writes and high contention pessimistic locking/semaphores would perform better.

Reply:
Regarding queue polling, that's a tricky part )
Ok, so whats going on. Here we have an underlying FIFO-queue containing every scheduled call associated with the moment it was added in the linked list. There is a predefined parameter `timeout` - every scheduled call should be spawned during such timeout, though it's not a hard deadline, just a desired property. And moreover, we want it to be as uniform over time as possible (in the presense of non-stationary process of new call arrivals).
On every polling iteration we repeatedly make a decision: pop and spawn next call from the queue or wait until the next polling iteraton. It'd be too expensive to obtain precise deadline distribution from the large queue, but since it's FIFO we know at least they are in the interval `[last_popped_deadline; timeout]`. And, of course, number of calls waiting in the queue is known as well.
Not much information, but we can use some randomized model for polling decisions and it may help to avoid synchronization with other queues. So here is the equation! It doesn't have solid theoretical underpinning, rather it's more an empirical one, reflecting two basic assumptions: probability of taking a call should be: 1) inversely proportional to the worst case deadline (when all calls in the queue share the same deadline) and 2) sublinear of a queue size.
So far its the best model I've been able to come up with in terms of smooth operation and meeting deadlines according to both our simulations and real app metrics. But I'll be happy to replace it with a more solid-ground algorithm, if any.

Reply:
Thanks for the explanation! Happy to defer to you here. We can play with it during Testground testing.

Reply:
(👋🏻 @dgtony really great context here—small suggestion would be to include it in a comment for future readers!)

Reply:
added comment and renamed vars for clarity